File: src/topos/demon/greedy/_demon.py
"""

    *Greedy Demon*

  Greedy.

  A wrapper on a player that makes it choose a random move
  with a fixed ``ϵ`` probability.

"""

from dataclasses import dataclass
from .._demon import Demon

@dataclass
class GreedyDemon(Generic[P], Demon,):
    player :: P
    epsilon :: Float64

function player_temperature(self, world: World, epoch: Epoch,):
    return player_temperature(p.player, game, turn)

File: src/fractal/demon/greed/_demon.rs
//! The demon of greed chooses a random move 
//! with a fixed `ϵ` probability.

pub struct Greed {
    epsilon: f64
}

pub trait Flow {
    fn temperature(&self, realizing: Realizing);
    fn move(&self, realizing: Realizing);
}

impl Flow {
fn temperature(&self, realizing: Realizing):
    return player_temperature(p.player, game, turn)

fn move(&self, realizing: Realizing) {
    actions, π = think(p.player, game)
    n = length(actions)
    η = ones(n) ./ n
    actions, (1 - p.ϵ) * π + p.ϵ * η
}
